src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs

```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Reflection.Metadata
+{
+    /// <summary>Specifies a type that should receive notifications of metadata updates.</summary>
+    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
+    public sealed class MetadataUpdateHandlerAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute.</summary>
+        /// <param name="handlerType">A type that handles metadata updates and that should be notified when any occur.</param>
+        public MetadataUpdateHandlerAttribute([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type handlerType) =>
```

Comment:
@terrajobst, I deviated slightly from what's in the API issue here.  The argument was just named "type", and there wasn't a property to get the value.  I renamed it to "handlerType" and added a "HandlerType" property.  Any concerns?

Reply:
I imagine we would want to trim the handlers and associated hot reload code in a published / trimmed application. Does adding the DynamicallyAccessedMembers attribute make that more difficult?

Reply:
> As long as we tell the linker it should trim the MetadataUpdateHandlerAttribute when publishing

Who is going to do that? If not here, can we log an issue to track it?

> then that unroots things it might use dynamically

Correct, if the only thing holding onto `RuntimeTypeMetadataUpdateHandler` is the attribute referencing it `[assembly: MetadataUpdateHandler(typeof(RuntimeTypeMetadataUpdateHandler))]`, then if the trimmer was instructed to trim all `MetadataUpdateHandler` attributes, it can also remove the `RuntimeTypeMetadataUpdateHandler` class.

Reply:
Where/how do I say the attribute should be trimmed if Debugger.IsSupported is false?

Reply:
Yes, you should be able to add an entry in that file, just like the existing ones.

> is that just about trimming these out of corelib, or does that end up applying to all assemblies?

The way it works is the `<assembly fullname="System.Private.CoreLib"` identifies which assembly the attribute is defined in. These entries tell the trimmer to remove the attribute definition and all instantiations of the attribute in all assemblies. However, note that if the trimmer sees code reading the attribute, the trimmer will leave the attributes in, and won't trim them. The reasoning is that if there is code in the app reading the attribute, then it shouldn't be trimmed.

An issue with some attributes (like `NullableContextAttribute`) is that they can be defined in any assembly, the attribute isn't just defined in one assembly. In that case, we need to say `<assembly fullname="*"`. But here, this attribute should only ever be defined in CoreLib.